omnipaxos_storage/src/memory.rs

```diff
@@ -1,239 +1,377 @@
-                snapshot: None,
-                stopsign: None,
-                marker: PhantomData::default(),
+                c_log,
+                c_path,
+                db,
+                t: PhantomData::default(),
+                s: PhantomData::default(),
             }
         }
-
-        // Create path and options for DB
-        pub fn optimize_storage(c_path: &str) -> (LogOptions, Options) {
-            let mut c_opts = LogOptions::new(&c_path);
-            c_opts.segment_max_bytes(0x10000); // 64 kB for each segment (entry)
-            c_opts.index_max_items(5000); // Max 10,000 log entries in the commitlog
-            c_opts.message_max_bytes(64000); // Max 64 kilobytes for each message
-
-            let mut db_opts = Options::default();
-            db_opts.set_write_buffer_size(0); // Set amount data to build up in single memtable before converting to on-disk file
-            db_opts.set_db_write_buffer_size(0); // Set total amount of data to build up in all memtables before writing to disk
-            db_opts.set_compression_type(rocksdb::DBCompressionType::Lz4); // set compression type
-            db_opts.set_max_write_buffer_number(1); // Max buffers for writing in memory
-            db_opts.set_max_write_buffer_size_to_maintain(0); // max size of write buffers to maintain in memory
-            db_opts.set_use_direct_reads(true); // data read from disk will not be cache or buffered
-            db_opts.set_use_direct_io_for_flush_and_compaction(true); // data flushed or compacted will not be cached or buffered
-            db_opts.increase_parallelism(4); // Set the amount threads for rocksDB compaction and flushing
-            db_opts.create_if_missing(true); // Creates an database if its missing in the path
-
-            (c_opts, db_opts)
-        }
     }
 
-    impl<T, S> Storage<T, S> for PersistentState<T, S>
+    impl<T, S> Storage<T, S> for PersistentStorage<T, S>
     where
-        T: Entry + AsBytes + FromBytes,
-        S: Snapshot<T>,
+        T: Entry + Serialize + for<'a> Deserialize<'a>,
+        S: Snapshot<T> + Serialize + for<'a> Deserialize<'a>,
     {
         fn append_entry(&mut self, entry: T) -> u64 {
-            //println!("append entry {:?}", entry);
-            let entry_bytes = AsBytes::as_bytes(&entry);
+            let entry_bytes = bincode::serialize(&entry).unwrap();
             match self.c_log.append_msg(entry_bytes) {
                 Ok(x) => x,
-                Err(_e) => 0,
+                Err(_e) => panic!(),
             }
         }
 
         fn append_entries(&mut self, entries: Vec<T>) -> u64 {
-            //println!("append entries!");
-            let mut buf: MessageBuf = MessageBuf::default();
             for entry in entries {
-                let _ = buf.push(AsBytes::as_bytes(&entry));
+                self.append_entry(entry);
             }
-            self.c_log.append(&mut buf).unwrap();
             self.get_log_len()
```

Comment:
This function can be done nicer I think.

Can use `let serialized = entries.into_iter().map()` to create a vec with all serialized entries. And then return `self.c_log.append(...) + 1` (if the +1 comment from `append_entry` is true)

Reply:
Fortunately `MessageBuf`, the only type that `commitlog.append` accepts, lets you create it with an iterator :+1: